lib/r10k/source/exec.rb

```diff
@@ -0,0 +1,53 @@
+require 'r10k/util/subprocess'
+require 'json'
+require 'yaml'
+
+class R10K::Source::Exec < R10K::Source::Hash
+  R10K::Source.register(:exec, self)
+
+  def initialize(name, basedir, options = {})
+    unless @command = options[:command]
+      raise ConfigError, _('Environment source %{name} missing required parameter: command') % {name: name}
+    end
+
+    # We haven't set the environments option yet. We will do that by
+    # overloading the #environments method.
+    super(name, basedir, options)
+  end
+
+  def environments
+    if @environments.nil?
+      set_environments_hash(run_environments_command)
+      super
+    end
+    @environments
+  end
+
+  def run_environments_command
+    subproc = R10K::Util::Subprocess.new([@command])
+    subproc.raise_on_fail = true
+    subproc.logger = self.logger
+    procresult = subproc.execute
+
+    begin
+      environments = JSON.parse(procresult.stdout)
+    rescue JSON::ParserError => json_err
+      begin
+        environments = YAML.load(procresult.stdout)
+      rescue Psych::SyntaxError => yaml_err
+        raise R10K::Error, _("Error parsing command output for exec source %{name}:\n" \
+                             "Not valid JSON: %{j_msg}\n" \
+                             "Not valid YAML: %{y_msg}\n" \
+                             "Stdout:\n%{out}") % {name: name, j_msg: json_err.message, y_msg: yaml_err.message, out: procresult.stdout}
+      end
+    end
+
+    unless R10K::Source::Hash.valid_environments_hash?(environments)
+      raise R10K::Error, _("Environment source ${name} command %{cmd} did not return valid environment data.\n" \
+                           'Returned: %{data}') % {name: name, cmd: command, data: environments}
```

Comment:
Blocker: the error handling in this code will fail if called.

1. The `${name}` variable expansion isn't a valid template expansion token
2. `@command` doesn't have an attr_reader so this code will throw an exception during error handling:

```suggestion
      raise R10K::Error, _("Environment source %{name} command %{cmd} did not return valid environment data.\n" \
                           'Returned: %{data}') % {name: name, cmd: @command, data: environments}
```

I've added a test for this as well:

```diff
diff --git i/lib/r10k/source/exec.rb w/lib/r10k/source/exec.rb
index 8460699..87a3a0e 100644
--- i/lib/r10k/source/exec.rb
+++ w/lib/r10k/source/exec.rb
@@ -43,8 +43,8 @@ class R10K::Source::Exec < R10K::Source::Hash
     end
 
     unless R10K::Source::Hash.valid_environments_hash?(environments)
-      raise R10K::Error, _("Environment source ${name} command %{cmd} did not return valid environment data.\n" \
-                           'Returned: %{data}') % {name: name, cmd: command, data: environments}
+      raise R10K::Error, _("Environment source %{name} command %{cmd} did not return valid environment data.\n" \
+                           'Returned: %{data}') % {name: name, cmd: @command, data: environments}
     end
 
     # Return the resulting environments hash
diff --git i/spec/unit/source/exec_spec.rb w/spec/unit/source/exec_spec.rb
index 51fc069..fa36f7c 100644
--- i/spec/unit/source/exec_spec.rb
+++ w/spec/unit/source/exec_spec.rb
@@ -66,6 +66,15 @@ describe R10K::Source::Exec do
           source = described_class.new('execsource', '/some/nonexistent/dir', command: '/path/to/command')
           expect { source.environments }.to raise_error(/Error parsing command output/)
         end
+
+        it 'raises an error for yaml data that is not a hash' do
+          allow_any_instance_of(R10K::Util::Subprocess)
+            .to receive(:execute)
+            .and_return(double('result', stdout: "[one, two]"))
+
+          source = described_class.new('execsource', '/some/nonexistent/dir', command: '/path/to/command')
+          expect { source.environments }.to raise_error(R10K::Error, /Environment source execsource.*did not return valid environment data.*one.*two/)
+        end
       end
     end
   end
```

Reply:
thanks for looking at this more closely than I did 😅